### ISSUE TITLE
Add lint checking for no-op uses of `Option::{as_deref,as_deref_mut}`

### ISSUE THREAD

**Issue by LeSeulArtichaut** [rust-lang/rust-clippy#7571]
### What it does

Checks for no-op uses of `Option::{as_deref,as_deref_mut}`.

### Categories (optional)

- Kind: `complexity`

*What is the advantage of the recommended code over the original code*

Removes useless code, removing clutter and improving readability.

### Drawbacks

None.

### Example

```rust
let a = Some(&1);
let b = a.as_deref(); // goes from Option<&i32> to Option<&i32>
```

Could be written as:

```rust
let a = Some(&1);
let b = a;
```

**Comment by camsteffen** [comment]
Should be easy as two `expr_ty` calls and check that they are the same type.

**Comment by lengyijun** [comment]
`Result<T,E>` `as_deref`  to  `Result<&<T as Deref>::Target, &E>`
Due to `&E` is not `E`, we shouldn't lint on `Result`.

**Comment by LeSeulArtichaut** [comment]
@lengyijun Correct, I missed that. Edited the issue
